In [22]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

## The ward Data

In [23]:
ward = pd.read_pickle('./dataset/ward.p')
print(ward.shape)
ward.head()

(50, 4)


ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

## The census Data

In [24]:
census = pd.read_pickle('./dataset/census.p')
print(census.shape)
census.head()

(50, 6)


ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

## Inner join

For clarity
- Tables == DataFrame
- merging == joining

<img src="./media/inner join.png">

In [25]:
ward_census = ward.merge(census, on='ward')
print(ward_census.shape)
ward_census.head()
#by default meger add suffixes of _x and _y on left and right data repectively

(50, 9)


ward            alderman                        address_x  zip_x  pop_2000  \
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647     52951   
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622     54361   
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609     40385   
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616     51953   
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649     55302   

   pop_2010 change                                address_y  zip_y  
0     56149     6%              2765 WEST SAINT MARY STREET  60647  
1     55805     3%                 WM WASTE MANAGEMENT 1500  60622  
2     53039    31%                      17 EAST 38TH STREET  60653  
3     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL  60653  
4     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE  60637

In [26]:
ward_census.columns

Index(['ward', 'alderman', 'address_x', 'zip_x', 'pop_2000', 'pop_2010',
       'change', 'address_y', 'zip_y'],
      dtype='object')

In [27]:
# to give custmoize suffix
ward_census = ward.merge(census, on='ward', suffixes=('_ward', '_cen'))
print(ward_census.shape)
ward_census.head()

(50, 9)


ward            alderman                     address_ward zip_ward  \
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE    60647   
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE    60622   
2    3          Pat Dowell          5046 SOUTH STATE STREET    60609   
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR    60616   
4    5  Leslie A. Hairston            2325 EAST 71ST STREET    60649   

   pop_2000  pop_2010 change                              address_cen zip_cen  
0     52951     56149     6%              2765 WEST SAINT MARY STREET   60647  
1     54361     55805     3%                 WM WASTE MANAGEMENT 1500   60622  
2     40385     53039    31%                      17 EAST 38TH STREET   60653  
3     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   60653  
4     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   60637

## Relationships : Types of relationships
- one to one
- one to many
- many to one 
- many to many

### one to one 
- Every row in left table is related to one rows in right table.
- Eg: Ward & Cences Example

### One to Many
- Every row in the left tabke is related to one or more rows in the right table
- Let's think back to our wards table. **Within each ward, there are many businesses**. We will merge the wards table with a table of ``licensed businesses`` in each ward.
- The **business license** data is stored in another table called ``licenses``. It holds info such as the business address and ward the business is located within.



In [28]:
licenses = pd.read_pickle('./dataset/licenses.p')
print(licenses.shape)
licenses.head()

(10000, 6)


account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

In [29]:
ward_licenses = ward.merge(licenses, on='ward', suffixes=('_ward', '_linc'))
print(ward_licenses.shape)
ward_licenses.head()

(10000, 9)


ward            alderman               address_ward zip_ward account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   12024  NaN   
1    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14446  743   
2    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14624  775   
3    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14987  NaN   
4    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   15642  814   

               business             address_linc zip_linc  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE    60622  
1      EMPTY BOTTLE INC   1035 N WESTERN AVE 1ST    60622  
2  LITTLE MEL'S HOT DOG    2205 N CALIFORNIA AVE    60647  
3    MR. BROWN'S LOUNGE   2301 W CHICAGO AVE 1ST    60622  
4          Beat Kitchen  2000-2100 W DIVISION ST    60622

- Notice that ward 1 and its alderman ``Joe`` is repeated in the resulting table because the **licenses table has many businesses** in the ``1st`` ward. 
- Pandas takes care of the one-to-many relationships for us and **doesn't require anything special** on our end. 
- We can use the same syntax as we did with **one-to-one** relationships.

In [30]:
#Note When you merge tables that have a one-to-many relationship, the number of rows returned will likely be different than the number in the left table
print(ward.shape)
print(ward_licenses.shape)

(50, 4)
(10000, 9)


## Merging multiple DataFrames

<img src= './media/multiple_dataframe.png'>

In [31]:
licenses.head()

account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

In [32]:
ward.head()

ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

In [33]:
# new dataset
grants = pd.read_pickle('./dataset/zip_demo.p')
print(grants.shape)
grants.head()

(66, 2)


zip  income
0  60630   70122
1  60640   50488
2  60622   87143
3  60614  100116
4  60608   41226

In [34]:
grants['business'] = 'REGGIE\'S BAR & GRILL'
grants

zip  income              business
0   60630   70122  REGGIE'S BAR & GRILL
1   60640   50488  REGGIE'S BAR & GRILL
2   60622   87143  REGGIE'S BAR & GRILL
3   60614  100116  REGGIE'S BAR & GRILL
4   60608   41226  REGGIE'S BAR & GRILL
..    ...     ...                   ...
61  53045  100438  REGGIE'S BAR & GRILL
62  60076   74964  REGGIE'S BAR & GRILL
63  60126  103454  REGGIE'S BAR & GRILL
64  60458   47411  REGGIE'S BAR & GRILL
65  60423  100098  REGGIE'S BAR & GRILL

[66 rows x 3 columns]

- If we merge the two tables only using the zip column, then the **60630 zip of Reggie's bar** from the licenses table will be matched to multiple businesses in the grants table with the same zip.
- The output of the **merge duplicates Reggie's** bar for each matching zip in the grants table, which is not what we want.

In [35]:
grants_licenses= grants.merge(licenses, on='zip')
grants_licenses.head()

zip  income            business_x account ward  aid  \
0  60630   70122  REGGIE'S BAR & GRILL   10548   39  NaN   
1  60630   70122  REGGIE'S BAR & GRILL   11390   35  NaN   
2  60630   70122  REGGIE'S BAR & GRILL     122   45  785   
3  60630   70122  REGGIE'S BAR & GRILL      14   39  763   
4  60630   70122  REGGIE'S BAR & GRILL   14572   45  775   

                     business_y                        address  
0  CORPORATE CLAIMS SYSTEMS INC              5247 N ELSTON AVE  
1          McGILL'S BAR & GRILL  4551-4553 N PULASKI RD GROUND  
2                  ZIEMEK CORP.           5368 N MILWAUKEE AVE  
3               KILDARE LIQUORS            4300 W LAWRENCE AVE  
4                SICILIA BAKERY            5939 W LAWRENCE AVE

## Single merge

In [36]:
ward.merge(licenses, on=['ward', 'zip']).head()

ward            alderman                  address_x    zip account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   14624  775   
1    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16004  NaN   
2    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16547  763   
3    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16662  763   
4    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16899  775   

                   business                   address_y  
0      LITTLE MEL'S HOT DOG       2205 N CALIFORNIA AVE  
1   EL NUEVO NARANJO LOUNGE    2210 N MILWAUKEE AVE 1ST  
2      PETE'S FOOD & LIQUOR     2556 W ARMITAGE AVE 1ST  
3  BUCK TOWN FOOD & LIQUORS        1950 N MILWAUKEE AVE  
4              JACK'S PLACE  2917 W ARMITAGE AVE  GROUN

In [37]:
ward.merge(licenses, on=['ward', 'zip'], suffixes=('_ward', '_linc')).head()

ward            alderman               address_ward    zip account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   14624  775   
1    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16004  NaN   
2    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16547  763   
3    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16662  763   
4    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16899  775   

                   business                address_linc  
0      LITTLE MEL'S HOT DOG       2205 N CALIFORNIA AVE  
1   EL NUEVO NARANJO LOUNGE    2210 N MILWAUKEE AVE 1ST  
2      PETE'S FOOD & LIQUOR     2556 W ARMITAGE AVE 1ST  
3  BUCK TOWN FOOD & LIQUORS        1950 N MILWAUKEE AVE  
4              JACK'S PLACE  2917 W ARMITAGE AVE  GROUN

## multiple merge

In [38]:
ward.merge(licenses, on=['ward', 'zip'], suffixes=('_ward', '_linc'))\
            .merge(grants, on= 'zip', suffixes=('_grant', '_wr_linc' )).head()

ward            alderman               address_ward    zip account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   14624  775   
1    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16004  NaN   
2    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16547  763   
3    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16662  763   
4    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE  60647   16899  775   

             business_grant                address_linc  income  \
0      LITTLE MEL'S HOT DOG       2205 N CALIFORNIA AVE   68223   
1   EL NUEVO NARANJO LOUNGE    2210 N MILWAUKEE AVE 1ST   68223   
2      PETE'S FOOD & LIQUOR     2556 W ARMITAGE AVE 1ST   68223   
3  BUCK TOWN FOOD & LIQUORS        1950 N MILWAUKEE AVE   68223   
4              JACK'S PLACE  2917 W ARMITAGE AVE  GROUN   68223   

       business_wr_linc  
0  REGGIE'S BAR & GRILL  
1  REGGIE'S BAR & GRILL  
2  REGGIE'S BAR & GRILL  
3  REGGIE'S BAR & GRILL  
4  REGGIE'S BAR & GRILL

## left join

<img src= './media/left_join.png' width="500" height="600">

<img src = './media/leftjoin_table.png' width="700" height="800">

### Dataset: The Movies DB

In [39]:
#movies Table
movies = pd.read_pickle('./dataset/movies/movies.p')
print(movies.shape)
movies.head()

(4803, 4)


id                 title  popularity release_date
0    257          Oliver Twist   20.415572   2005-09-23
1  14290  Better Luck Tomorrow    3.877036   2002-01-12
2  38365             Grown Ups   38.864027   2010-06-24
3   9672              Infamous    3.680896   2006-11-16
4  12819       Alpha and Omega   12.300789   2010-09-17

In [40]:
#tagline table
taglines = pd.read_pickle('./dataset/movies/taglines.p')
print(taglines.shape)
taglines.head()

(3955, 2)


id                                         tagline
0   19995                     Enter the World of Pandora.
1     285  At the end of the world, the adventure begins.
2  206647                           A Plan No One Escapes
3   49026                                 The Legend Ends
4   49529            Lost in our world, found in another.

In [41]:
#Merge left join
movies.merge(taglines, on='id', how='left')

id                 title  popularity release_date  \
0       257          Oliver Twist   20.415572   2005-09-23   
1     14290  Better Luck Tomorrow    3.877036   2002-01-12   
2     38365             Grown Ups   38.864027   2010-06-24   
3      9672              Infamous    3.680896   2006-11-16   
4     12819       Alpha and Omega   12.300789   2010-09-17   
...     ...                   ...         ...          ...   
4798   3089             Red River    5.344815   1948-08-26   
4799  11934   The Hudsucker Proxy   14.188982   1994-03-11   
4800  13807                Exiled    8.486390   2006-09-06   
4801  73873          Albert Nobbs    7.802245   2011-12-21   
4802  11622   Blast from the Past    8.737058   1999-02-12   

                                                tagline  
0                                                   NaN  
1                  Never underestimate an overachiever.  
2       Boys will be boys. . . some longer than others.  
3               There's more to the story than you know  
4                                A Pawsome 3D Adventure  
...                                                 ...  
4798  Big as the men who faced this challenge! Bold ...  
4799  They took him for a fall guy... but he threw t...  
4800                                                NaN  
4801         A man with a secret. A woman with a dream.  
4802  She'd never met anyone like him. He's never me...  

[4803 rows x 5 columns]

<img src='./media/right_join.png' width=500 height=600>

<img src='./media/rightjoin_table.png' width=700 height=800>

In [98]:
movies_to_genre = pd.read_pickle('./dataset/movies/movie_to_genres.p')
print(movies_to_genre.shape)
movies_to_genre.head()

(12160, 2)


movie_id            genre
0         5            Crime
1         5           Comedy
2        11  Science Fiction
3        11           Action
4        11        Adventure

In [43]:
#filtering the data
# condition = movies_to_genre['genre'] == 'TV Movie'
#tv_genre = movies_to_genre[condition]
tv_genre = movies_to_genre[movies_to_genre['genre'] == 'TV Movie']
tv_genre

movie_id     genre
4998      10947  TV Movie
5994      13187  TV Movie
7443      22488  TV Movie
10061     78814  TV Movie
10790    153397  TV Movie
10835    158150  TV Movie
11096    205321  TV Movie
11282    231617  TV Movie

In [44]:
# right join
# movies.merge(tv_genre, on='movie_id', how='right') #<-- Error because didn't match columns

In [45]:
# right join
movies.merge(tv_genre, how='right', left_on='id', right_on='movie_id') #<-- No Error because columns are defined

id                      title  popularity release_date  movie_id  \
0   10947        High School Musical   16.536374   2006-01-20     10947   
1   13187  A Charlie Brown Christmas    8.701183   1965-12-09     13187   
2   22488         Love's Abiding Joy    1.128559   2006-10-06     22488   
3   78814       We Have Your Husband    0.102003   2011-11-12     78814   
4  153397                   Restless    0.812776   2012-12-07    153397   
5  158150        How to Fall in Love    1.923514   2012-07-21    158150   
6  205321                  Sharknado   20.466433   2013-07-11    205321   
7  231617  Signed, Sealed, Delivered    1.444476   2013-10-13    231617   

      genre  
0  TV Movie  
1  TV Movie  
2  TV Movie  
3  TV Movie  
4  TV Movie  
5  TV Movie  
6  TV Movie  
7  TV Movie

- Our returned table has movies that match our table of tv_genres. - There **does not** appear to be any null values in the columns from the movies table. 

<img src='./media/outer_join.png' width=500 height=600>

<img src='./media/outerjoin_table.png' width=700 height=800>

In [46]:
family = movies_to_genre[movies_to_genre['genre']== 'Family']
f = family.reset_index(drop=True).head(3)
f

movie_id   genre
0        12  Family
1        35  Family
2       105  Family

In [47]:
comedy = movies_to_genre[movies_to_genre['genre']== 'Comedy']
c = comedy.reset_index(drop=True).head(3)
c

movie_id   genre
0         5  Comedy
1        13  Comedy
2        35  Comedy

In [48]:
family_comedy = f.merge(c, on='movie_id', how='outer', suffixes=('_fam','_com'))
family_comedy

movie_id genre_fam genre_com
0        12    Family       NaN
1        35    Family    Comedy
2       105    Family       NaN
3         5       NaN    Comedy
4        13       NaN    Comedy

## for merging more tables

<img src='./media/moretables.png'>

### Assignment # 04 : Apply all the four joins that you had learned.

## Merging a table to itself

<img src='./media/mergingtabletoitself.png'>

In [102]:
sequel = pd.read_pickle('./dataset/movies/sequels.p')
print(sequel.shape)
sequel.head()

(4803, 3)


id         title  sequel
0  19995        Avatar    <NA>
1    862     Toy Story     863
2    863   Toy Story 2   10193
3    597       Titanic    <NA>
4  24428  The Avengers    <NA>

In [50]:
original_sequels = sequel.merge(sequel, left_on='sequel', right_on='id', suffixes=('_org', '_seq'))
original_sequels

id_org                                          title_org  sequel_org  \
0      862                                          Toy Story         863   
1      863                                        Toy Story 2       10193   
2      675          Harry Potter and the Order of the Phoenix         767   
3      121              The Lord of the Rings: The Two Towers         122   
4      120  The Lord of the Rings: The Fellowship of the Ring         121   
..     ...                                                ...         ...   
85      76                                     Before Sunrise          80   
86    2292                                             Clerks        2295   
87    9367                                        El Mariachi        8068   
88    8374                                The Boondock Saints       22821   
89   16186                         Diary of a Mad Black Woman       16781   

    id_seq                                      title_seq  sequel_seq  
0      863                                    Toy Story 2       10193  
1    10193                                    Toy Story 3        <NA>  
2      767         Harry Potter and the Half-Blood Prince        <NA>  
3      122  The Lord of the Rings: The Return of the King        <NA>  
4      121          The Lord of the Rings: The Two Towers         122  
..     ...                                            ...         ...  
85      80                                  Before Sunset      132344  
86    2295                                      Clerks II        <NA>  
87    8068                                      Desperado        1428  
88   22821         The Boondock Saints II: All Saints Day        <NA>  
89   16781                         Madea's Family Reunion       15670  

[90 rows x 6 columns]

In [51]:
original_sequels['title_org'], original_sequels['title_seq'].head()

(0                                             Toy Story
 1                                           Toy Story 2
 2             Harry Potter and the Order of the Phoenix
 3                 The Lord of the Rings: The Two Towers
 4     The Lord of the Rings: The Fellowship of the Ring
                             ...                        
 85                                       Before Sunrise
 86                                               Clerks
 87                                          El Mariachi
 88                                  The Boondock Saints
 89                           Diary of a Mad Black Woman
 Name: title_org, Length: 90, dtype: object,
 0                                      Toy Story 2
 1                                      Toy Story 3
 2           Harry Potter and the Half-Blood Prince
 3    The Lord of the Rings: The Return of the King
 4            The Lord of the Rings: The Two Towers
 Name: title_seq, dtype: object)

In [52]:
original_sequels[['title_org', 'title_seq']].head()

title_org  \
0                                          Toy Story   
1                                        Toy Story 2   
2          Harry Potter and the Order of the Phoenix   
3              The Lord of the Rings: The Two Towers   
4  The Lord of the Rings: The Fellowship of the Ring   

                                       title_seq  
0                                    Toy Story 2  
1                                    Toy Story 3  
2         Harry Potter and the Half-Blood Prince  
3  The Lord of the Rings: The Return of the King  
4          The Lord of the Rings: The Two Towers

### Merging table to itself with left join

In [53]:
original_sequels = sequel.merge(sequel,
                                left_on='sequel',
                                right_on='id',
                                how='left',
                               suffixes=('_org','_seq'))
original_sequels.head()

id_org     title_org  sequel_org  id_seq    title_seq  sequel_seq
0   19995        Avatar        <NA>    <NA>          NaN        <NA>
1     862     Toy Story         863     863  Toy Story 2       10193
2     863   Toy Story 2       10193   10193  Toy Story 3        <NA>
3     597       Titanic        <NA>    <NA>          NaN        <NA>
4   24428  The Avengers        <NA>    <NA>          NaN        <NA>

### Merging table to itself with right join

In [54]:
original_sequels = sequel.merge(sequel,
                                left_on='sequel',
                                right_on='id',
                                how='right',
                               suffixes=('_org','_seq'))
original_sequels.head()

id_org  title_org  sequel_org  id_seq     title_seq  sequel_seq
0    <NA>        NaN        <NA>   19995        Avatar        <NA>
1    <NA>        NaN        <NA>     862     Toy Story         863
2     862  Toy Story         863     863   Toy Story 2       10193
3    <NA>        NaN        <NA>     597       Titanic        <NA>
4    <NA>        NaN        <NA>   24428  The Avengers        <NA>

### When to merge at table to itself
Common situations:
- Hierarchical relationships eg: employee & manager
- Sequential relationships eg: logistic movements 
- Graph data eg: network of friends

## Merging on Indexes

In [55]:
#setting an index
new_movie = movies.set_index('id')
new_movie

title  popularity release_date
id                                                  
257            Oliver Twist   20.415572   2005-09-23
14290  Better Luck Tomorrow    3.877036   2002-01-12
38365             Grown Ups   38.864027   2010-06-24
9672               Infamous    3.680896   2006-11-16
12819       Alpha and Omega   12.300789   2010-09-17
...                     ...         ...          ...
3089              Red River    5.344815   1948-08-26
11934   The Hudsucker Proxy   14.188982   1994-03-11
13807                Exiled    8.486390   2006-09-06
73873          Albert Nobbs    7.802245   2011-12-21
11622   Blast from the Past    8.737058   1999-02-12

[4803 rows x 3 columns]

In [56]:
taglines

id                                            tagline
0      19995                        Enter the World of Pandora.
1        285     At the end of the world, the adventure begins.
2     206647                              A Plan No One Escapes
3      49026                                    The Legend Ends
4      49529               Lost in our world, found in another.
...      ...                                                ...
4795  124606            Sometimes you've got to break the rules
4796   14337                 What happens if it actually works?
4798    9367  He didn't come looking for trouble, but troubl...
4799   72766  A newlywed couple's honeymoon is upended by th...
4801  126186                           A New Yorker in Shanghai

[3955 rows x 2 columns]

In [57]:
new_movie.merge(taglines, on='id', how='left').head()

id                 title  popularity release_date  \
0    257          Oliver Twist   20.415572   2005-09-23   
1  14290  Better Luck Tomorrow    3.877036   2002-01-12   
2  38365             Grown Ups   38.864027   2010-06-24   
3   9672              Infamous    3.680896   2006-11-16   
4  12819       Alpha and Omega   12.300789   2010-09-17   

                                           tagline  
0                                              NaN  
1             Never underestimate an overachiever.  
2  Boys will be boys. . . some longer than others.  
3          There's more to the story than you know  
4                           A Pawsome 3D Adventure

### multiple index

In [58]:
main_casts = pd.read_pickle('./dataset/movies/casts.p')
print(main_casts.shape)
main_casts.head()

(106257, 6)


movie_id  cast_id character  gender    id               name
7          5       22   Jezebel       1  3122        Sammi Davis
8          5       23     Diana       1  3123  Amanda de Cadenet
9          5       24    Athena       1  3124     Valeria Golino
3          5       25   Elspeth       1  3125            Madonna
12         5       26       Eva       1  3126          Ione Skye

In [59]:
main_casts = main_casts.set_index(['movie_id', 'cast_id'])
main_casts

character  gender       id               name
movie_id cast_id                                              
5        22        Jezebel       1     3122        Sammi Davis
         23          Diana       1     3123  Amanda de Cadenet
         24         Athena       1     3124     Valeria Golino
         25        Elspeth       1     3125            Madonna
         26            Eva       1     3126          Ione Skye
...                    ...     ...      ...                ...
433715   3           Amber       0  1500111      Nicole Smolen
         4              BB       0  1734573        Kim Baldwin
         5           Sugar       0  1734574    Ariana Stephens
         6            Drew       0  1734575        Bryson Funk
459488   0        Narrator       0  1354401    Tony Oppedisano

[106257 rows x 4 columns]

In [60]:
casts = main_casts['character']
casts = pd.DataFrame(casts)
casts

character
movie_id cast_id          
5        22        Jezebel
         23          Diana
         24         Athena
         25        Elspeth
         26            Eva
...                    ...
433715   3           Amber
         4              BB
         5           Sugar
         6            Drew
459488   0        Narrator

[106257 rows x 1 columns]

In [61]:
samuel = main_casts[main_casts.name == 'Samuel L. Jackson']
samuel = pd.DataFrame(samuel['name'])
samuel.head()

name
movie_id cast_id                   
184      3        Samuel L. Jackson
319      13       Samuel L. Jackson
326      2        Samuel L. Jackson
329      138      Samuel L. Jackson
393      21       Samuel L. Jackson

In [62]:
samuel_casts = samuel.merge(casts, on=['movie_id', 'cast_id'])
print(samuel_casts.shape)
samuel_casts.head()

(67, 2)


name      character
movie_id cast_id                                  
184      3        Samuel L. Jackson  Ordell Robbie
319      13       Samuel L. Jackson        Big Don
326      2        Samuel L. Jackson  Neville Flynn
329      138      Samuel L. Jackson         Arnold
393      21       Samuel L. Jackson          Rufus

## Index merge with left_on and right_on

In [63]:
movies = movies.set_index('id')
movies_to_genre = movies_to_genre.set_index('movie_id')

In [67]:
movies_genre = movies.merge(movies_to_genre, left_on='id', right_on='movie_id', right_index=True)
movies_genre.head()

title  popularity release_date   genre
id                                                          
257            Oliver Twist   20.415572   2005-09-23   Crime
257            Oliver Twist   20.415572   2005-09-23   Drama
257            Oliver Twist   20.415572   2005-09-23  Family
14290  Better Luck Tomorrow    3.877036   2002-01-12   Crime
14290  Better Luck Tomorrow    3.877036   2002-01-12   Drama

## Practise

- Merge the movies table, as the left table, with the financials table using a left join, and save the result to movies_financials.
- Count the number of rows in movies_financials with a null value in the budget column.
- print number_of_missing_fin

In [68]:
movies = pd.read_pickle('./dataset/movies/movies.p')
financials = pd.read_pickle('./dataset/movies/financials.p')

In [69]:
# Merge the movies table with the financials table with a left join
movies_financials = movies.merge(financials, how='left', on='id')
movies_financials.head()

id                 title  popularity release_date      budget  \
0    257          Oliver Twist   20.415572   2005-09-23  50000000.0   
1  14290  Better Luck Tomorrow    3.877036   2002-01-12         NaN   
2  38365             Grown Ups   38.864027   2010-06-24  80000000.0   
3   9672              Infamous    3.680896   2006-11-16  13000000.0   
4  12819       Alpha and Omega   12.300789   2010-09-17  20000000.0   

       revenue  
0   42093706.0  
1          NaN  
2  271430189.0  
3    1151330.0  
4   39300000.0

In [70]:
# Count the number of rows in the budget column that are missing
number_of_missing_fin = movies_financials['budget'].isnull().sum()

In [71]:
# Print the number of movies missing financials
print(number_of_missing_fin)

1574


- Merge toy_story and taglines on the id column with a left join, and save the result as toystory_tag.
- With toy_story as the left table, merge to it taglines on the id column with an inner join, and save as toystory_tag.

In [72]:
toy_story = pd.read_pickle('./dataset/movies/sequels.p')

In [76]:
# Merge the toy_story and taglines tables with a left join
toystory_tag = toy_story.merge(taglines, on='id', how='left')

# Print the rows and shape of toystory_tag
print(toystory_tag.shape)
toystory_tag.head()

(4803, 4)


id         title  sequel                                    tagline
0  19995        Avatar    <NA>                Enter the World of Pandora.
1    862     Toy Story     863                                        NaN
2    863   Toy Story 2   10193                         The toys are back!
3    597       Titanic    <NA>  Nothing on Earth could come between them.
4  24428  The Avengers    <NA>                                        NaN

In [78]:
# Merge the toy_story and taglines tables with a inner join
toystory_tag = toy_story.merge(taglines, on='id')

# Print the rows and shape of toystory_tag
print(toystory_tag.shape)
toystory_tag.head()

(3955, 4)


id           title  sequel                                    tagline
0   19995          Avatar    <NA>                Enter the World of Pandora.
1     863     Toy Story 2   10193                         The toys are back!
2     597         Titanic    <NA>  Nothing on Earth could come between them.
3  135397  Jurassic World    <NA>                          The park is open.
4  168259       Furious 7    <NA>                        Vengeance Hits Home

- Merge action_movies and scifi_movies tables with a right join on movie_id. Save the result as action_scifi.
- Update the merge to add suffixes, where '_act' and '_sci' are suffixes for the left and right tables, respectively.
- From action_scifi, subset only the rows where the genre_act column is null.
- Merge movies and scifi_only using the id column in the left table and the movie_id column in the right table with an inner join.

In [88]:
action_movies = movies_to_genre[movies_to_genre['genre']== 'Action']
action_movies.reset_index().head()

movie_id   genre
0        11  Action
1        18  Action
2        22  Action
3        24  Action
4        58  Action

In [89]:
scifi_movies = movies_to_genre[movies_to_genre['genre']== 'Science Fiction']
scifi_movies.reset_index().head()

movie_id            genre
0        11  Science Fiction
1        18  Science Fiction
2        19  Science Fiction
3        38  Science Fiction
4        62  Science Fiction

In [93]:
# Merge action_movies to the scifi_movies with right join
action_scifi = action_movies.merge(scifi_movies, on='movie_id', how='right',
                                   suffixes=('_act','_sci'))

# From action_scifi, select only the rows where the genre_act column is null
scifi_only = action_scifi[action_scifi['genre_act'].isnull()]

# Merge the movies and scifi_only tables with an inner join
movies_and_scifi_only = movies.merge(scifi_only, left_on='id', right_on='movie_id')

# Print the first few rows and shape of movies_and_scifi_only
print(movies_and_scifi_only.shape)
movies_and_scifi_only.head()

(258, 6)


id                         title  popularity release_date genre_act  \
0  18841  The Lost Skeleton of Cadavra    1.680525   2001-09-12       NaN   
1  26672     The Thief and the Cobbler    2.439184   1993-09-23       NaN   
2  15301      Twilight Zone: The Movie   12.902975   1983-06-24       NaN   
3   8452                   The 6th Day   18.447479   2000-11-17       NaN   
4   1649    Bill & Ted's Bogus Journey   11.349664   1991-07-19       NaN   

         genre_sci  
0  Science Fiction  
1  Science Fiction  
2  Science Fiction  
3  Science Fiction  
4  Science Fiction

- Merge movie_to_genres and pop_movies using a right join. Save the results as genres_movies.
- Group genres_movies by genre and count the number of id values.

In [ ]:
# Use right join to merge the movie_to_genres and pop_movies tables
genres_movies = movie_to_genres.merge(pop_movies, how='right', 
                                      left_on='movie_id', 
                                      right_on= 'id')

# Count the number of genres
genre_count = genres_movies.groupby('genre').agg({'id':'count'})

# Plot a bar chart of the genre_count
genre_count.plot(kind='bar')
plt.show()

- To a variable called crews_self_merged, merge the crews table to itself on the id column using an inner join, setting the suffixes to '_dir' and '_crew' for the left and right tables respectively.
- Create a Boolean index, named boolean_filter, that selects rows from the left table with the job of 'Director' and avoids rows with the job of 'Director' in the right table.
- Use the .head() method to print the first few rows of direct_crews.

In [ ]:
# Merge the crews table to itself
crews_self_merged = crews.merge(crews, on='id', how='inner',
                                suffixes=('_dir','_crew'))

# Create a boolean index to select the appropriate rows
boolean_filter = ((crews_self_merged['job_dir'] == 'Director') & 
                  (crews_self_merged['job_crew'] != 'Director'))
direct_crews = crews_self_merged[boolean_filter]

# Print the first few rows of direct_crews
print(direct_crews.head())

- With the sequels table on the left, merge to it the financials table on index named id, ensuring that all the rows from the sequels are returned and some rows from the other table may not be returned, Save the results to sequels_fin.
- Merge the sequels_fin table to itself with an inner join, where the left and right tables merge on sequel and id respectively with suffixes equal to ('_org','_seq'), saving to orig_seq.
- Select the title_org, title_seq, and diff columns of orig_seq and save this as titles_diff.
- Sort by titles_diff by diff in descending order and print the first few rows.

In [ ]:
# Merge sequels and financials on index id
sequels_fin = sequels.merge(financials, on='id', how='left')

# Self merge with suffixes as inner join with left on sequel and right on id
orig_seq = sequels_fin.merge(sequels_fin, how='inner', left_on='sequel', 
                             right_on='id', right_index=True,
                             suffixes=('_org','_seq'))

# Add calculation to subtract revenue_org from revenue_seq 
orig_seq['diff'] = orig_seq['revenue_seq'] - orig_seq['revenue_org']

# Select the title_org, title_seq, and diff 
titles_diff = orig_seq[['title_org','title_seq','diff']]

# Print the first rows of the sorted titles_diff
print(titles_diff.sort_values('diff', ascending=False).head())